### PR TITLE
Spell check support for treesitter

### PR DESCRIPTION
**Pull request by lewis6991**
#16620 is a prerequisite which adds `vim.spell.check`

Here's a bold attempt to add spellchecking support to the built in highlighter.

This basically works via a new `spellcheck!` directive which creates "sub_captures" in the metadata containing ranges within the captured node. Sub capture names are dot appended to the main capture.

Example query for lua would be:

```
(
   (comment) @spell
   (#spellcheck! @spell)
)
```

Which would create sub-captures: `spell.bad`, `spell.rare`, `spell.local` and `spell.caps`. These would be mapped onto highlight groups in nvim-treesitter.

Questions:
- Do we even want to do this in core? Do we just keep https://github.com/lewis6991/spellsitter.nvim as a separate plugin?
- Is doing this via a directive a good approach?

Note:
We will need to disable vim-syntax spellchecking but with `set spell` enabled. The only way I've found to do this is via a nasty hack:
https://github.com/lewis6991/spellsitter.nvim/blob/504e89725ec68b163f1a59294031a43b950836dc/lua/spellsitter.lua#L260-L267

This create an empty syntax item called `Spell` and prevents Neovim running the builtin spellchecker.

### CODE REVIEW

**Comment by lewis6991** [comment]
@vigoux @gpanders

**Comment by vigoux** [comment]
So, the `sub_capture` thing looks pretty not right to me for now, but I can't see a neat way of avoiding it.
Another way of doing it would be through a special `@Spell` capture, and handle all the spellchecking on the C / extmarks side (kinda like what was done with extmarks beforehad).

I would really like @bfredl's opinion on that.

Another thing is that I think it clobbers the highlighting part of tree-sitter with this, which is technically not quite highlighting...

**Comment by gpanders** [comment]
> Do we even want to do this in core?

Big +1 from me, we already have spell checking in core, why not leverage treesitter to do it?

**Comment by theHamsta** [comment]
Wouldn't it be better to handle spelling independently from highlighting: e.g. have `query/<lang>/spell.scm` and have the spell module attach to the language tree (it already supports registering into its callbacks) instead of integrating the spell code into the module.

**Comment by bfredl** [comment]
If we do this in core, I really think we should hook into the builtin checking logic in win_line.c, so we now it matches exactly and we don't need to maintain a repeat implementation. I think an ephemeral extmark should be able to override if a region in a buffer is spellchecked or not (just like syntax.c logic can decide when spell.c is called into in win_line). It could be an attribute on the Decoration object.

**Comment by lewis6991** [comment]
> Wouldn't it be better to handle spelling independently from highlighting: e.g. have `query/<lang>/spell.scm` and have the spell module attach to the language tree (it already supports registering into its callbacks) instead of integrating the spell code into the module.

This is exactly what spellsitter already does. The problem is spell checking really needs to be done in the redraw events, whereas the treesitter callbacks are done in `buf_attach` callbacks which is not what we want. Spellsitter is basically I trimmed down version of `vim.treesitter.highlighter` and does its own tree traversal. The point of this PR is to remove that extra tree traversal and do it all together to be more efficient.

**Comment by lewis6991** [comment]
> If we do this in core, I really think we should hook into the builtin checking logic in win_line.c, so we know it matches exactly and we don't need to maintain a repeat implementation. I think an ephemeral extmark should be able to override if a region in a buffer is spellchecked or not (just like syntax.c logic can decide when spell.c is called into in win_line). It could be an attribute on the Decoration object.

So if I understand correctly, you basically want something of the form:

```lua
a.nvim_buf_set_extmark(buf, ns, start_row, start_col,
                       { end_line = end_row, end_col = end_col,
                         hl_group = capture == 'spell' and  {'SpellBad', 'SpellRare', 'SpellLocal', 'SpellCaps' } or hl,
                         ephemeral = true,
                         priority = tonumber(metadata.priority), or 100 -- Low but leaves room below
                         spell = capture == 'spell'
                        })
```

Where, if `spell=true`, would only highlight the spelling errors with the highlights passed to it? (assuming we don't want to use hard-coded highlights like we do now)

The only benefit I can see with this is that it's maybe a little faster?

Two drawbacks here:
1. We would have to handle spelling specially in the highlighter as opposed to keeping it generic.
2. It locks users into using spell.c's `spell_check()` as a spelling provider. With this PR anyone can override the `spellcheck!` directive or even `vim.spell.check` to use any provider they want.

**Comment by theHamsta** [comment]
>  The point of this PR is to remove that extra tree traversal and do it all together to be more efficient.

I'm just worried that we're mixing different features here. Is running two queries so much slower than running the union of the two. The highlighter code is already super messy and adding more code doesn't make it better. Probably we want to do something similar for conceal later.

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 1]
```diff
@@ -253,15 +258,33 @@ local function on_line_impl(self, buf, line)
...
+          if hl and end_row >= line then
+            a.nvim_buf_set_extmark(buf, ns, start_row, start_col,
+                                   { end_line = end_row, end_col = end_col,
+                                     hl_group = hl,
+                                     ephemeral = true,
+                                     priority = tonumber(metadata.priority) or 100 -- Low but leaves room below
+                                    })
+          end
+        end
+      else
```
Comment (theHamsta) [inline runtime/lua/vim/treesitter/highlighter.lua:278]: I'm having problems to understand this else here, so there will be no normal highlight and only the highlights of the subcaptures when there are subcaptures?

Comment (lewis6991) [reply runtime/lua/vim/treesitter/highlighter.lua:278]: Yeah I guess it makes sense to always do this branch. I'll change.

**Comment by bfredl** [comment]
>   We would have to handle spelling specially in the highlighter as opposed to keeping it generic.

Que? The solution I presented is (slightly) more generic, anything that provides extmarks can toggle spellcheck behavior. "handle spelling specially in the highlighter" is something this PR does already, so how is that a drawback when considering an alternative?

>    It locks users into using spell.c's spell_check() as a spelling provider. With this PR anyone can override the spellcheck! directive or even vim.spell.check to use any provider they want.

Except that it doesn't "lock in" the user in any way at all. I am not sure what this is even supposed to mean. The addition/extension of a core API by definition does not prevent something to be implemented in user space, if it could be done before.

**Comment by lewis6991** [comment]
> Que? The solution I presented is (slightly) more generic, anything that provides extmarks can toggle spellcheck behavior. "handle spelling specially in the highlighter" is something this PR does already, so how is that a drawback when considering an alternative?

Well currently the highlighter is completely agnostic to the names of any captures or highlights (they are all managed in nvim-treesitter), and as I showed in my example we would need to then pass in the spell highlights explicitly; again assuming we don't want to use hard-coded highlights and maintain the same framework we use for all other TS highlights. And even if we do use hard-coded highlights, we would probably still need to reference the new special `@spell` capture in the highlighter directly. Maybe you could generalize it by having a special sub-category of captures just for spelling which nvim-treesitter can populate?

And I'd say it is quite a bit more generic. It's a mechanism for operating/refining captures via a directive (which is effectively a callback in ways), and an alternative to full language injection. E.g if https://github.com/stsewd/tree-sitter-comment didn't exist, we could use this mechanism to achieve the same thing by just adding a simple directive and an equivalent query. How about highlighting issue references only in comments? Or doing something like https://github.com/norcalli/nvim-colorizer.lua but only in CSS field values? I don't know, it's just an idea.

> Except that it doesn't "lock in" the user in any way at all. I am not sure what this is even supposed to mean. The addition/extension of a core API by definition does not prevent something to be implemented in user space, if it could be done before.

OK sure, It can be done currently, like spellsitter does, however you would need to implement a separate highlighter in addition to any logic for it to play nice with the existing spell interface (mainly `set spell`). All I'm pointing out is that one nice benefit of this PR (with #16620) is that it leaves mechanisms to redefine **only** the spell checker provider via a simple interface. Maybe we can make the extmark logic look for other providers and not only use `spell_check()`?

I'm honestly just trying to weigh the pros/cons of each approach here.

**Comment by bfredl** [comment]
> And I'd say it is quite a bit more generic. It's a mechanism for operating/refining captures via a directive (which is effectively a callback in ways), and an alternative to full language injection. 

Sure but now we are changing the question to something else to what I, in fact, responded to ( what this PR actually claims to do and how it does it). Enabling the builtin syntax highlighting from tree-sitter does not preclude to use a generalized `sub_capture` mechanism for other things.

> OK sure, It can be done currently, like spellsitter does, however you would need to implement a separate highlighter in addition to any logic for it to play nice with the existing spell interface (mainly set spell).

Sure but one reason to go via the established syntax interface is that it does not just provide highlights, but also navigation commands (though the latter is gonna require persistent marks to be efficient). I think we would like to enable separate spell checker providers in a way that **integrates** with the full UX associated with the existing mechanism, rather than having third-party providers of checking needing to also re-implement the find/correct interface from scratch.

**Pull request closed without merging**